### PR TITLE
Fix duplicate meta descriptions causing generic SEO snippets

### PR DESCRIPTION
**Pull request by chenders**
## Summary

- **Root cause**: `index.html` had hardcoded `<meta name="description">` and `<title>` tags that remained in SSR output alongside Helmet-injected page-specific tags at `<!--app-head-->`, producing duplicate tags
- **Impact**: Google was using the generic description ("Look up any movie and see which actors have passed away...") for every page instead of the unique per-page descriptions already generated by React Helmet
- **Fix**: Removed hardcoded tags from the template so SSR provides them exclusively via Helmet; SPA fallback now injects defaults at the placeholder

## Test plan

- [x] All SEO component tests pass (10 tests)
- [x] All SSR middleware tests pass (16 tests)
- [x] All prerender middleware tests pass (16 tests)
- [x] All prerender renderer tests pass (16 tests)
- [x] Page component tests pass (HomePage, ActorPage, MoviePage — 64 tests)
- [x] Vite production build succeeds
- [x] Built `index.html` confirmed to have no hardcoded `<meta name="description">` or `<title>`
- [ ] After deploy: flush SSR Redis cache (`ssr:*` keys), then verify unique descriptions via `curl -s <url> | grep '<meta name="description"'`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: server/src/middleware/ssr.ts [hunk 1]
```diff
@@ -259,14 +259,22 @@ function assembleHtml(
   return html
 }
 
+/** Default head tags injected when SSR is unavailable (SPA fallback). */
+const DEFAULT_HEAD_TAGS = [
+  "<title>Dead on Film - Movie Cast Mortality Database</title>",
+  '<meta name="description" content="Look up any movie and see which actors have passed away. Discover mortality statistics, death dates, and causes of death for your favorite films." />',
+].join("\n    ")
+
```
[the remark this answers is not in the document]
Comment (chenders) [reply server/src/middleware/ssr.ts]: Fixed in f64f8c9b. `assembleHtml()` now always replaces `<!--app-head-->`, falling back to `DEFAULT_HEAD_TAGS` when `headTags` is empty. This ensures every SSR response has a title and description, and prevents the placeholder comment from leaking.

File: server/src/middleware/ssr.ts [hunk 2]
```diff
@@ -259,14 +259,22 @@ function assembleHtml(
...
  */
 function serveSpaFallback(res: Response): void {
   try {
     const template = getTemplate()
-    // Strip SSR placeholders — client will do a full render
-    const fallback = template.replace("<!--app-head-->", "").replace("<!--app-html-->", "")
+    // Inject default title + description, then strip app placeholder
+    const fallback = template
+      .replace("<!--app-head-->", DEFAULT_HEAD_TAGS)
+      .replace("<!--app-html-->", "")
```
[the remark this answers is not in the document]
Comment (chenders) [reply server/src/middleware/ssr.ts:284]: Fixed in f64f8c9b. The SPA fallback test now asserts the default title and description are present in the rendered HTML, and that the `<!--app-head-->` placeholder is not present.

File: server/src/middleware/ssr.ts [hunk 3]
```diff
@@ -259,14 +258,22 @@ function assembleHtml(
   return html
 }
 
+/** Default head tags injected when SSR is unavailable (SPA fallback). */
+const DEFAULT_HEAD_TAGS = [
+  "<title>Dead on Film - Movie Cast Mortality Database</title>",
+  '<meta name="description" content="Look up any movie and see which actors have passed away. Discover mortality statistics, death dates, and causes of death for your favorite films." />',
+].join("\n    ")
```
[the remark this answers is not in the document]
Comment (chenders) [reply server/src/middleware/ssr.ts]: Fixed in 9b86630b. Updated docstring to reflect both SPA fallback and SSR empty-Helmet usage.

File: server/src/lib/prerender/data-fetchers.ts [hunk 4]
```diff
@@ -47,6 +47,20 @@ function extractYear(date: string | Date | null | undefined): string | null {
...
 
+/** Calculate current age from a birthday string (e.g., "1945-07-26"). */
+function calculateCurrentAge(birthday: string | Date | null | undefined): number | null {
+  if (!birthday) return null
+  const birth = birthday instanceof Date ? birthday : new Date(birthday)
+  if (Number.isNaN(birth.getTime())) return null
+  const today = new Date()
+  let age = today.getFullYear() - birth.getFullYear()
+  const monthDiff = today.getMonth() - birth.getMonth()
+  if (monthDiff < 0 || (monthDiff === 0 && today.getDate() < birth.getDate())) {
```
[the remark this answers is not in the document]
Comment (chenders) [reply server/src/lib/prerender/data-fetchers.ts]: Fixed in 9b86630b. Switched to UTC getters (getUTCFullYear/getUTCMonth/getUTCDate) to avoid timezone-dependent off-by-one errors.

File: server/src/lib/prerender/data-fetchers.ts [hunk 5]
```diff
@@ -227,15 +241,19 @@ async function getActorPageData(actorId: number): Promise<PrerenderPageData | nu
...
+  // Build a rich description matching the client-side Helmet output
+  let description: string
+  if (isDeceased) {
+    const deathYear = extractYear(actor.deathday)
+    const agePart = actor.age_at_death ? ` at age ${actor.age_at_death}` : ""
+    const causePart = actor.cause_of_death ? ` Cause of death: ${actor.cause_of_death}.` : ""
+    description = `${actor.name} died in ${deathYear || "unknown"}${agePart}.${causePart} See complete filmography and mortality statistics.`
+  } else {
+    const currentAge = calculateCurrentAge(actor.birthday)
+    const agePart = currentAge ? ` at age ${currentAge}` : ""
```
[the remark this answers is not in the document]
Comment (chenders) [reply server/src/lib/prerender/data-fetchers.ts]: Fixed in 9b86630b. Changed to `!= null` checks for both `age_at_death` and `currentAge`.

File: server/src/lib/prerender/data-fetchers.ts [hunk 6]
```diff
@@ -227,15 +241,19 @@ async function getActorPageData(actorId: number): Promise<PrerenderPageData | nu
...
+    description = `${actor.name} is alive${agePart}. See filmography and which co-stars have passed away.`
+  }
```
[the remark this answers is not in the document]
Comment (chenders) [reply server/src/lib/prerender/data-fetchers.ts]: Fixed in 9b86630b. Extracted `buildActorDescription()` as an exported helper and added 7 tests covering: deceased with/without cause, with/without age, living with/without birthday, age 0 edge case, and Date objects from pg driver.

File: index.html [hunk 7]
```diff
@@ -12,7 +12,6 @@
     <meta name="theme-color" content="#3d2914" />
     <meta name="theme-color" content="#141210" media="(prefers-color-scheme: dark)" />
     <meta name="viewport" content="width=device-width, initial-scale=1.0" />
-    <meta name="description" content="Look up any movie and see which actors have passed away. Discover mortality statistics, death dates, and causes of death for your favorite films." />
     <!-- Preload self-hosted fonts for faster rendering -->
     <link rel="preload" href="/fonts/inter-var.woff2" as="font" type="font/woff2" crossorigin>
```
[the remark this answers is not in the document]
Comment (chenders) [reply index.html:16]: Fixed in 6a7e7784. Restored default <title> in index.html for SPA users. Both assembleHtml() and serveSpaFallback() now strip the template title before injecting Helmet/default tags, so SSR output has exactly one <title>. The meta description stays removed (that was the actual SEO bug). Test verifies no duplicate titles.

File: server/src/middleware/ssr.ts [hunk 8]
```diff
@@ -239,10 +239,9 @@ function assembleHtml(
 ): string {
   let html = template
 
-  // Inject head tags from Helmet
-  if (headTags) {
-    html = html.replace("<!--app-head-->", headTags)
-  }
+  // Inject head tags from Helmet, falling back to defaults so every response
+  // has a <title> and meta description even if Helmet produced nothing.
+  html = html.replace("<!--app-head-->", headTags || DEFAULT_HEAD_TAGS)
```
[the remark this answers is not in the document]
Comment (chenders) [reply server/src/middleware/ssr.ts:251]: Fixed in 6a7e7784. The SPA fallback test now asserts exactly one <title> tag in output (no duplicates), confirming the template title is stripped when DEFAULT_HEAD_TAGS is injected. The mock template was also updated to include a <title> matching the real index.html, making the test realistic for the title-stripping logic.

File: server/src/middleware/ssr.ts [hunk 9]
```diff
@@ -239,10 +239,14 @@ function assembleHtml(
...
 
-  // Inject head tags from Helmet
-  if (headTags) {
-    html = html.replace("<!--app-head-->", headTags)
+  // Inject head tags from Helmet, falling back to defaults so every response
+  // has a <title> and meta description even if Helmet produced nothing.
+  const injectedTags = headTags || DEFAULT_HEAD_TAGS
+  // Strip the template's default <title> to avoid duplicates when Helmet provides one
+  if (injectedTags.includes("<title>")) {
+    html = html.replace(/<title>[^<]*<\/title>/, "")
```
[the remark this answers is not in the document]
Comment (chenders) [reply server/src/middleware/ssr.ts]: Fixed in b1a067c7. Changed to /<title\b/i regex for detection and /<title[^>]*>[^<]*<\/title>/i for stripping, handling Helmet output with attributes like data-rh. Also updated the same pattern in serveSpaFallback.
